utils/src/main/java/org/apache/cloudstack/utils/process/ProcessRunner.java

```diff
@@ -0,0 +1,112 @@
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class ProcessRunner {
+    public static final Logger LOG = Logger.getLogger(ProcessRunner.class);
+
+    private static final ExecutorService processExecutor = Executors.newCachedThreadPool(new NamedThreadFactory("ProcessRunner"));
+
+    private static String readStream(final InputStream inputStream) {
+        String text = null;
+        try {
+            final BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
+            String line;
+            while ((line = bufferedReader.readLine()) != null) {
+                if (Strings.isNullOrEmpty(text)) {
+                    text = line;
+                } else {
+                    text = text + "\n" + line;
+                }
+            }
+        } catch (IOException e) {
+            if (LOG.isTraceEnabled()) {
+                LOG.trace("ProcessRunner::readStream failed due to: " + e.getMessage());
+            }
+        }
+        return text;
+    }
+
+    public static ProcessResult executeCommands(final List<String> commands, final Duration timeOut) {
+        Preconditions.checkArgument(commands != null && timeOut != null);
+
+        int retVal = -2;
+        String stdOutput = null;
+        String stdError = null;
+
+        try {
+            final Process process = new ProcessBuilder().command(commands).start();
+            if (timeOut.getStandardSeconds() > 0) {
+                final Future<Integer> processFuture = processExecutor.submit(new Callable<Integer>() {
+                    @Override
+                    public Integer call() throws Exception {
+                        return process.waitFor();
+                    }
+                });
+                try {
+                    retVal = processFuture.get(timeOut.getStandardSeconds(), TimeUnit.SECONDS);
+                } catch (ExecutionException | TimeoutException e) {
+                    retVal = -1;
+                    stdError = "Operation timed out, aborted";
+                    if (LOG.isTraceEnabled()) {
+                        LOG.trace("Failed to complete the requested command within timeout: " + e.getMessage());
+                    }
+                } finally {
+                    if (Strings.isNullOrEmpty(stdError)) {
+                        stdOutput = readStream(process.getInputStream());
+                        stdError = readStream(process.getErrorStream());
+                    }
+                }
+            } else {
+                retVal = process.waitFor();
+                stdOutput = readStream(process.getInputStream());
+                stdError = readStream(process.getErrorStream());
+            }
+            process.destroy();
```

Comment:
process.destroy() is common to both cases when timeout is 0 or non-zero; cleanup is ensured in either case, as all exceptions are handled.